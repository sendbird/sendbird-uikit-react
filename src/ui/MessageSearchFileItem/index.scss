@import '../../styles/variables';

.sendbird-message-search-file-item {
  position: relative;
  display: inline-flex;
  flex-direction: row;
  width: 100%;
  height: 80px;
  cursor: pointer;
  @include themed() {
    border-bottom: 1px solid t(on-bg-4);
    background-color: t(bg-0);
  }

  &:hover {
    @include themed() {
      background-color: t(bg-1);
    }
  }

  .sendbird-message-search-file-item__left {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 90px;
    height: 80px;
  }
  .sendbird-message-search-file-item__right {
    position: relative;
    display: flex;
    flex-direction: column;
    width: calc(100% - 106px);

    .sendbird-message-search-file-item__right__sender-name {
      position: absolute;
      top: 12px;
      display: flex;
      max-width: 280px;
      height: 16px;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .sendbird-message-search-file-item__right__content {
      position: absolute;
      top: 36px;
      box-sizing: border-box;
      width: 100%;
      height: 32px;
      display: flex;
      flex-direction: row;
      align-items: center;

      .sendbird-message-search-file-item__right__content__type-icon {
        position: relative;
        margin-right: 4px;
        display: flex;
        align-items: center;
        justify-content: center;
        width: 26px;
        min-width: 26px;
        height: 26px;
        border-radius: 8px;
        @include themed() {
          background-color: t(bg-1);
        }
      }
      .sendbird-message-search-file-item__right__content__url {
        height: 16px;
        position: relative;
        display: inline;
        overflow: hidden;
        text-overflow: ellipsis;
        word-break: keep-all;
        white-space: nowrap;
      }
    }
  }
  .sendbird-message-search-file-item__message-created-at {
    position: absolute;
    top: 12px;
    right: 18px;
  }
  .sendbird-message-search-file-item__right-footer {
    width: 16px;
  }
}

.sendbird-message-search-file-item.sendbird-message-search-file-item--selected {
  @include themed() {
    background-color: t(primary--1-bg--7);
  }
  .sendbird-message-search-file-item__left {
    @include themed() {
      border-left: 4px solid t(primary-3);
    }
  }
}
