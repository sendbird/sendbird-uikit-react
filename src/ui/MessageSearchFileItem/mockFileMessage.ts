export const docMock = {
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1579767478896,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.test',
  name: '',
  size: 0,
  type: '',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: 'hoon1234',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const imageMock = {
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1579767478896,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.jpg',
  name: '',
  size: 0,
  type: '',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: 'hoon1234',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const audioMock = {
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1579767478896,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp3',
  name: '',
  size: 0,
  type: '',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: 'hoon1234',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const videoMock = {
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1579767478896,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp4',
  name: '',
  size: 0,
  type: '',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: 'hoon1234',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const gifMock = {
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1579767478896,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.gif',
  name: '',
  size: 0,
  type: '',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: 'hoon1234',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

