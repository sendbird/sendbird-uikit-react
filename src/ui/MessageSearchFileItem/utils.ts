import type { FileMessage } from '@sendbird/chat/message';
import type { Locale } from 'date-fns';
import format from 'date-fns/format';
import isToday from 'date-fns/isToday';
import isThisYear from 'date-fns/isThisYear';
import isYesterday from 'date-fns/isYesterday';
import { IconTypes } from '../Icon';
import { isVoiceMessage } from '../../utils';
import { getMessageFirstFileUrl } from '../QuoteMessage/utils';
import { MultipleFilesMessage } from '@sendbird/chat/message';
import { Types } from '../Icon/type';

export interface GetCreatedAtProps {
  createdAt: number;
  locale?: Locale;
  stringSet?: Record<string, string>;
}

export function getCreatedAt({ createdAt, locale, stringSet }: GetCreatedAtProps): string {
  const optionalParam = locale ? { locale } : {};
  if (!createdAt) {
    return '';
  }
  if (isToday(createdAt)) {
    return format(createdAt, 'p', optionalParam);
  }
  if (isYesterday(createdAt)) {
    return stringSet?.MESSAGE_STATUS__YESTERDAY || 'Yesterday';
  }
  if (isThisYear(createdAt)) {
    return format(createdAt, 'MMM d', optionalParam);
  }
  return format(createdAt, 'yyyy/M/d', optionalParam);
}

export function getIconOfFileType(message: FileMessage | MultipleFilesMessage): Types {
  const fileMessageUrl = getMessageFirstFileUrl(message) ?? '';
  const fileExtension = (fileMessageUrl.match(/\.([^.]*?)(?=\?|#|$)/))?.[1] ?? '';

  if (/(jpg|jpeg|png)$/i.test(fileExtension)) {
    return IconTypes.PHOTO;
  } else if (/mp4$/i.test(fileExtension) || isVoiceMessage(message)) {
    return IconTypes.PLAY;
  } else if (/mp3/i.test(fileExtension)) {
    return IconTypes.FILE_AUDIO;
  } else if (/gif/i.test(fileExtension)) {
    return IconTypes.GIF;
  } else {
    return IconTypes.FILE_DOCUMENT;
  }
}

export function truncate(fullText: string, textLimit: number): string {
  if (fullText.length <= textLimit) return fullText;
  const separator = '...';
  const sepLen = separator.length;
  const charsToShow = textLimit - sepLen;
  const frontChars = Math.ceil(charsToShow / 2);
  const backChars = Math.floor(charsToShow / 2);
  return fullText.substr(0, frontChars) + separator + fullText.substr(fullText.length - backChars);
}
