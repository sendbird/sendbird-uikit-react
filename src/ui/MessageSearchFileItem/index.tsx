import './index.scss';
import React, { ReactElement } from 'react';
import type { FileMessage } from '@sendbird/chat/message';

import Avatar from '../Avatar';
import Icon, { IconColors } from '../Icon';
import Label, { LabelTypography, LabelColors } from '../Label';
import { useLocalization } from '../../lib/LocalizationContext';
import { getCreatedAt, getIconOfFileType } from './utils';
import { isVoiceMessage } from '../../utils';
import { MultipleFilesMessage } from '@sendbird/chat/message';
import { getMessageFirstFileName, getMessageFirstFileUrl } from '../QuoteMessage/utils';

interface Props {
  className?: string | Array<string>;
  message: FileMessage | MultipleFilesMessage;
  selected?: boolean;
  onClick?: (message: FileMessage | MultipleFilesMessage) => void;
}

export default function MessageSearchFileItem(props: Props): ReactElement {
  const {
    className,
    message,
    selected,
    onClick,
  } = props;
  const { createdAt } = message;
  const url: string = getMessageFirstFileUrl(message);
  const name: string = getMessageFirstFileName(message);
  // @ts-ignore
  const sender = message.sender || message._sender;
  const { profileUrl, nickname } = sender;
  const { stringSet, dateLocale } = useLocalization();
  const isVoiceMsg = isVoiceMessage(message);
  const prettyFilename = isVoiceMsg ? stringSet.VOICE_MESSAGE : (name || url);

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-message-search-file-item',
        selected ? 'sendbird-message-search-file-item--selected' : '',
      ].join(' ')}
      onClick={(e) => {
        e.stopPropagation();
        onClick?.(message);
      }}
    >
      <div className="sendbird-message-search-file-item__left">
        <Avatar
          className="sendbird-message-search-file-item__left__sender-avatar"
          src={profileUrl}
          alt="profile image"
          width="56px"
          height="56px"
        />
      </div>
      <div className="sendbird-message-search-file-item__right">
        <Label
          className="sendbird-message-search-file-item__right__sender-name"
          type={LabelTypography.SUBTITLE_2}
          color={LabelColors.ONBACKGROUND_1}
        >
          {nickname || stringSet.NO_NAME}
        </Label>
        <div className="sendbird-message-search-file-item__right__content">
          {!isVoiceMsg && (
            <div className={'sendbird-message-search-file-item__right__content__type-icon'}>
              <Icon
                type={getIconOfFileType(message)}
                fillColor={IconColors.PRIMARY}
                width="18px"
                height="18px"
              />
            </div>
          )}
          <Label
            className="sendbird-message-search-file-item__right__content__url"
            type={LabelTypography.BODY_2}
            color={LabelColors.ONBACKGROUND_1}
          >
            {prettyFilename}
          </Label>
        </div>
      </div>
      <Label
        className="sendbird-message-search-file-item__message-created-at"
        type={LabelTypography.CAPTION_3}
        color={LabelColors.ONBACKGROUND_2}
      >
        {getCreatedAt({ createdAt, locale: dateLocale, stringSet })}
      </Label>
      <div className="sendbird-message-search-file-item__right-footer" />
    </div>
  );
}
