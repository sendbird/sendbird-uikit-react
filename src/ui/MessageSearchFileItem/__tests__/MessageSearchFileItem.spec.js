import React from 'react';
import { render } from '@testing-library/react';

import MessageSearchFileItem from "../index";
import {
  docMock,
  // gifMock,
  // videoMock,
  // audioMock,
  // imageMock,
} from '../mockFileMessage';

jest.mock('date-fns/format', () => () => ('mock-date'));


describe('ui/MessageSearchFileItem', () => {
  // should add test cases for each file types
  // define id for each icon svg files first
  // https://sendbird.atlassian.net/browse/UK-634

  it('should do a snapshot test of the MessageSearchFileItem DOM', function() {
    const text = "example-text";
    const { asFragment } = render(
      <MessageSearchFileItem
        className={text}
        message={docMock}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
