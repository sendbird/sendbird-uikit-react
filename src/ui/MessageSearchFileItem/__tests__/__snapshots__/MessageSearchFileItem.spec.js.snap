// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/MessageSearchFileItem should do a snapshot test of the MessageSearchFileItem DOM 1`] = `
<DocumentFragment>
  <div
    class="example-text sendbird-message-search-file-item "
  >
    <div
      class="sendbird-message-search-file-item__left"
    >
      <div
        class="sendbird-message-search-file-item__left__sender-avatar sendbird-avatar"
        role="button"
        style="height: 56px; width: 56px; z-index: 0; left: 0px;"
        tabindex="0"
      >
        <div
          class="sendbird-avatar-img sendbird-image-renderer"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
        >
          <div
            class="sendbird-image-renderer__image"
            style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://static.sendbird.com/sample/profiles/profile_19_512px.png);"
          />
          <img
            alt="profile image"
            class="sendbird-image-renderer__hidden-image-loader"
            src="https://static.sendbird.com/sample/profiles/profile_19_512px.png"
          />
        </div>
      </div>
    </div>
    <div
      class="sendbird-message-search-file-item__right"
    >
      <span
        class="sendbird-message-search-file-item__right__sender-name sendbird-label sendbird-label--subtitle-2 sendbird-label--color-onbackground-1"
      >
        ME
      </span>
      <div
        class="sendbird-message-search-file-item__right__content"
      >
        <div
          class="sendbird-message-search-file-item__right__content__type-icon"
        >
          <div
            class=" sendbird-icon sendbird-icon-file-document sendbird-icon-color--primary"
            role="button"
            style="width: 18px; min-width: 18px; height: 18px; min-height: 18px;"
            tabindex="0"
          >
            <test-file-stub />
          </div>
        </div>
        <span
          class="sendbird-message-search-file-item__right__content__url sendbird-label sendbird-label--body-2 sendbird-label--color-onbackground-1"
        >
          72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.test
        </span>
      </div>
    </div>
    <span
      class="sendbird-message-search-file-item__message-created-at sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-2"
    >
      2020/1/23
    </span>
    <div
      class="sendbird-message-search-file-item__right-footer"
    />
  </div>
</DocumentFragment>
`;
